pkg/controllers/user/logging/utils/secret.go

```diff
@@ -0,0 +1,114 @@
+		loggingconfig.ClientCertName: []byte(""),
+		loggingconfig.ClientKeyName:  []byte(""),
+	}
+)
+
+func UpdateSSLAuthentication(prefix, target string, esConfig *v3.ElasticsearchConfig, spConfig *v3.SplunkConfig, kfConfig *v3.KafkaConfig, secrets rv1.SecretInterface) error {
+	var certificate, clientCert, clientKey string
+	if target == loggingconfig.Elasticsearch {
+		certificate = esConfig.Certificate
+		clientCert = esConfig.ClientCert
+		clientKey = esConfig.ClientKey
+	} else if target == loggingconfig.Splunk {
+		certificate = spConfig.Certificate
+		clientCert = spConfig.ClientCert
+		clientKey = spConfig.ClientKey
+	} else if target == loggingconfig.Kafka {
+		certificate = kfConfig.Certificate
+		clientCert = kfConfig.ClientCert
+		clientKey = kfConfig.ClientKey
+	}
+	return updateSecret(loggingconfig.SSLSecretName, prefix, certificate, clientCert, clientKey, secrets)
+}
+
+func InitSecret(secrets rv1.SecretInterface) error {
+	_, err := secrets.Controller().Lister().Get(loggingconfig.LoggingNamespace, loggingconfig.SSLSecretName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if _, err := secrets.Create(newSecret(loggingconfig.LoggingNamespace, loggingconfig.SSLSecretName, initSecretData)); err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	return nil
+}
+
+func UnsetSecret(secrets rv1.SecretInterface, name string) error {
+	ncm, err := secrets.Get(name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	ncm.Data = initSecretData
+
+	if _, err := secrets.Update(ncm); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+func updateSecret(name, prefix, certificate, clientCert, clientKey string, secrets rv1.SecretInterface) error {
+	existSecret, err := secrets.Get(name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			data := map[string][]byte{
+				prefix + "_" + loggingconfig.CaFileName:     []byte(certificate),
+				prefix + "_" + loggingconfig.ClientCertName: []byte(clientCert),
+				prefix + "_" + loggingconfig.ClientKeyName:  []byte(clientKey),
+			}
+
+			if existSecret, err = secrets.Create(newSecret(loggingconfig.LoggingNamespace, name, data)); err != nil && !apierrors.IsAlreadyExists(err) {
+				return errors.Wrapf(err, "create secret %s:%s failed", loggingconfig.LoggingNamespace, name)
+			}
+		} else {
+			return errors.Wrapf(err, "get secret %s:%s failed", loggingconfig.LoggingNamespace, name)
+		}
+		return errors.Wrap(err, "get secret failed")
+	}
+
+	newSecret := existSecret.DeepCopy()
+	newSecret.Data[prefix+"_"+loggingconfig.CaFileName] = []byte(certificate)
+	newSecret.Data[prefix+"_"+loggingconfig.ClientCertName] = []byte(clientCert)
+	newSecret.Data[prefix+"_"+loggingconfig.ClientKeyName] = []byte(clientKey)
+
+	if reflect.DeepEqual(existSecret.Data, newSecret.Data) {
+		return nil
+	}
+
+	if _, err = secrets.Update(newSecret); err != nil {
```

Comment:
is the secret ever being removed? I guess there are no implications of it being left around, but still

Reply:
add reset when related crd is removed, the secret is deployed in cattle-logging namespace, when there is no target config, the namespace will be deleted.